File: employee/employee-service/src/models/employee.ts
import mongoose from "mongoose";
import { EmployeeAttrs, EmployeeDoc, EmployeeModel } from "./interfaces";
import { updateIfCurrentPlugin } from "mongoose-update-if-current";
import { PersonMsg, GenderType, RaceTypes, ShiftPreference, SignInTypes, UserRoles } from "@espressotrip-org/concept-common";

/**
 * User model that uses update-if-current version incrementation
 */
const employeeSchema = new mongoose.Schema(
    {
        firstName: { type: String, required: true },
        lastName: { type: String, required: true },
        gender: { type: String, required: true },
        race: { type: String, required: true },
        position: { type: String, required: true },
        startDate: { type: String, required: true },
        shiftPreference: { type: String, required: true },
        branchName: { type: String, required: true },
        region: { type: String, required: true },
        registeredEmployee: { type: Boolean, default: true },
        division: { type: String, required: true },
        authId: { type: String, default: "" },
        country: { type: String, required: true },
        phoneNumber: { type: String, required: true },
        email: { type: String, required: true, unique: true },
        signInType: { type: String, required: true, default: SignInTypes.UNKNOWN },
        userRole: { type: String, required: true },
        password: { type: String, required: true },
        providerId: { type: String, default: "" },
    },
    {
        toJSON: {
            virtuals: true,
            transform(doc, ret) {
                ret.id = ret._id;
                delete ret._id;
            },
        },
    }
);

/** Replace the __v with version  && use the update-if-current plugin*/
employeeSchema.set("versionKey", "version");
employeeSchema.plugin(updateIfCurrentPlugin);

/** Static schema functions */
/**
 * Static function to build product
 * @param attributes
 */
employeeSchema.statics.build = function (attributes: EmployeeAttrs): EmployeeDoc {
    return new Employee(attributes);
};

/**
 * Converts employee document to employee message object to create a auth user
 * @param document {EmployeeDoc}
 * @param withVersion {boolean}
 */
employeeSchema.statics.convertToMessage = function (document: EmployeeDoc, withVersion: boolean): PersonMsg {
    const msg: PersonMsg = {
        id: document.id,
        country: document.country,
        email: document.email,
        gender: document.gender as GenderType,
        branchName: document.branchName,
        lastName: document.lastName,
        password: document.password,
        race: document.race as RaceTypes,
        region: document.region,
        registeredEmployee: document.registeredEmployee,
        division: document.division,
        position: document.position,
        providerId: document.providerId,
        shiftPreference: document.shiftPreference as ShiftPreference,
        startDate: document.startDate,
        signInType: document.signInType,
        userRole: document.userRole as UserRoles,
        firstName: document.firstName,
        phoneNumber: document.phoneNumber,
        version: document.version,
    };
    if (!withVersion) delete msg.version;
    return msg;
};

employeeSchema.statics.findByEvent = async function (employee: PersonMsg): Promise<EmployeeDoc | null> {
    if (!employee.version) return await Employee.findOne({ email: employee.email });
    return await Employee.findOne({ email: employee.email, version: employee.version - 1 });
};
/** Create model from schema */
const Employee = mongoose.model<EmployeeDoc, EmployeeModel>("Employee", employeeSchema);
export { Employee };

File: employee-dash/employee-dash-service/models/employee.go
package models

import (
	"go.mongodb.org/mongo-driver/bson/primitive"
	"time"
)

// Employee is the data structure used by the task service
type Employee struct {
	Id                 primitive.ObjectID `bson:"_id"`
	Country            string             `bson:"country"`
	Email              string             `bson:"email"`
	Gender             string             `bson:"gender"`
	BranchName         string             `bson:"branchName"`
	LastName           string             `bson:"lastName"`
	Password           string             `bson:"password"`
	Race               string             `bson:"race"`
	Region             string             `bson:"region"`
	RegisteredEmployee bool               `bson:"registeredEmployee"`
	Position           string             `bson:"position"`
	ProviderId         string             `bson:"providerId"`
	ShiftPreference    string             `bson:"shiftPreference"`
	StartDate          time.Time          `bson:"startDate"`
	SignInType         string             `bson:"signInType"`
	UserRole           string             `bson:"userRole"`
	FirstName          string             `bson:"firstName"`
	PhoneNumber        string             `bson:"phoneNumber"`
	Division           string             `bson:"division"`
	Version            int                `bson:"version"`
}


// EmployeePayload is the payload received from the Rabbit publishers
type EmployeePayload struct {
	Id                 string    `json:"id"`
	Country            string    `json:"country"`
	Email              string    `json:"email"`
	Gender             string    `json:"gender"`
	BranchName         string    `json:"branchName"`
	LastName           string    `json:"lastName"`
	Password           string    `json:"password"`
	Race               string    `json:"race"`
	Region             string    `json:"region"`
	RegisteredEmployee bool      `json:"registeredEmployee"`
	Position           string    `json:"position"`
	ProviderId         string    `json:"providerId"`
	ShiftPreference    string    `json:"shiftPreference"`
	StartDate          time.Time `json:"startDate"`
	SignInType         string    `json:"signInType"`
	UserRole           string    `json:"userRole"`
	FirstName          string    `json:"firstName"`
	PhoneNumber        string    `json:"phoneNumber"`
	Division           string    `json:"division"`
	Version            int       `json:"version"`
}
